sw/device/lib/dif/dif_kmac_unittest.cc

```diff
@@ -86,5 +90,90 @@ TEST(FunctionNameTest, BadArg) {
   EXPECT_EQ(dif_kmac_function_name_init("", -1, &fn), kDifBadArg);
 }
 
+using mock_mmio::MmioTest;
+using mock_mmio::MockDevice;
+using testing::Test;
+
+// Base class for the rest fixtures in this file.
+class KmacTest : public testing::Test, public mock_mmio::MmioTest {
+ public:
+  dif_kmac_t kmac_;
+  dif_kmac_operation_state_t op_state_;
+  const uint8_t kMsg_[17] = {0xa7, 0x48, 0x47, 0x93, 0x0a, 0x03,
+                             0xab, 0xee, 0xa4, 0x73, 0xe1, 0xf3,
+                             0xdc, 0x30, 0xb8, 0x88, 0x15};
+  const size_t kLen_ = sizeof(kMsg_);
+
+  KmacTest() { EXPECT_EQ(dif_kmac_init(dev().region(), &kmac_), kDifOk); }
+
+  /**
+   * Set mmio write expectation for 8 bits data size.
+   *
+   * @param message Buffer with the data.
+   * @param size Len of the buffer.
+   */
+  void setExpectedMessageByte(const uint8_t *message, const size_t size = 8) {
+    for (int i = 0; i < size; ++i) {
+      EXPECT_WRITE8(KMAC_MSG_FIFO_REG_OFFSET, message[i]);
+    }
+  }
+
+  /**
+   * Set mmio write expectation for 32 bits data size considering an alignment
+   * of 32 bits.
+   *
+   * @param message Buffer with the data.
+   * @param size Len of the buffer.
+   */
+  void setExpectedMessageInt32(const uint8_t *message, const size_t size = 8) {
+    // Check if the buffer is unaligned.
+    size_t remaining = size;
+    size_t unalignment = ((uintptr_t)message) % sizeof(uint32_t);
+    if (unalignment) {
+      // write unaligned data in bytes.
+      setExpectedMessageByte(message, unalignment);
+      message += unalignment;
+      remaining -= unalignment;
+    }
+
+    // Write aligned part of the buffer.
+    while (remaining > 3) {
+      uint32_t word = 0;
+      memcpy(&word, message, sizeof(uint32_t));
+      EXPECT_WRITE32(KMAC_MSG_FIFO_REG_OFFSET, word);
+      remaining -= sizeof(uint32_t);
+      message += sizeof(uint32_t);
+    }
+    // Check if there still unaligned data in the buffer tail.
+    if (remaining > 0) {
+      // write unaligned data in bytes.
+      setExpectedMessageByte(message, remaining);
+    }
+  }
+};
+
+TEST_F(KmacTest, AbsorbAlignedMessage) {
+  __attribute__((aligned(sizeof(uint32_t)))) uint8_t alignedMsg[sizeof(kMsg_)];
+  memcpy(alignedMsg, kMsg_, kLen_);
+
+  EXPECT_READ32(KMAC_STATUS_REG_OFFSET, 3);
+  setExpectedMessageInt32(alignedMsg, kLen_);
+
+  EXPECT_EQ(dif_kmac_absorb(&kmac_, &op_state_, alignedMsg, kLen_, NULL),
+            kDifOk);
+}
+
+TEST_F(KmacTest, AbsorbUnalignedMessage) {
+  __attribute__((aligned(sizeof(uint16_t))))
```

Comment:
This doesn't guarantee an unaligned input. Since this test is running on the host and execution time is not a concern it might be better to just test all relevant alignments by brute force:

```c
uint8_t buffer[kLen_ + 4];
for (int i = 0; i < 3; ++i) {
  uint8_t *unalignedMsg = &buffer[i];
  ...
}
```

This method would also mean you'd only need one test - since it would be testing both aligned and unaligned inputs in one go.

Reply:
good point that is